### PR TITLE
Swift: add CFG for normal autoclosures

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

File: swift/ql/lib/codeql/swift/controlflow/internal/ControlFlowElements.qll [hunk 1]
```diff
@@ -41,6 +41,13 @@ predicate isPropertyGetterElement(PropertyGetterElement pge, Accessor accessor,
   pge = TPropertyGetterElement(accessor, ref)
 }
 
+predicate isNormalAutoClosureOrExplicitClosure(ClosureExpr clos) {
+  clos instanceof AutoClosureExpr and
+  exists(CallExpr call | call.getAnArgument().getExpr() = clos)
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/controlflow/internal/ControlFlowElements.qll:46]: Why do we check that the `AutoClosureExpr` in an argument to a call?  When wouldn't it be?

Comment (rdmarsh2) [reply swift/ql/lib/codeql/swift/controlflow/internal/ControlFlowElements.qll:46]: Short-circuiting logical operators and nil-coalescing use autoclosure exprs also.

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/controlflow/internal/ControlFlowElements.qll:46]: Makes sense, thanks.

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/controlflow/internal/ControlFlowElements.qll:46]: To be completely clear for future readers, logical operators _are_ functions and they do have _arguments_, but crucially for this code they are not called by a `CallExpr`.  They are "called" by a `BinaryExpr` and that's the case we're excluding here.

**Review by geoffw0 — COMMENTED** [review]
I think the qldoc for a few of these classes could do with enhancing, so that this stuff can be easier to understand:
- `AutoclosureExpr` claims to model `@autoclosure` arguments (case 1), but it might be worth explictly mentioning that this includes the second arguments to `&&`, `||`, `??` (case 2).
- `ExplicitclosureExpr` has no qldoc, but I assume it means when you write a closure with `{}` (case 3).
- `TClosureElement` and `isNormalAutoClosureOrExplicitClosure` both apply in cases 1 and 3, but not 2.

---

And here's my understanding of the control flow classes:
- `ExplicitClosureTree` models case 3.
- `ShortCircuitingAutoClosureTree` models case 2.
- `AutoClosureTree` models case 1.
- `ClosureExprTree` models the `TClosureElement` / `ClosureElement` in cases 1 and 3 (previously just case 3).
	- I'm not entirely sure what the difference between the `TClosureElement` dataflow node and the existing dataflow node for the `ClosureExpr` expression is?  Is this one of these cases where we need an extra node for some transformation step?
- `ClosureExprScope` defines the range for the library, it now covers cases 1 and 3 (previously just 3)
	- I'm not entirely clear what this is really for or why it doesn't include all of the above.

---

I'm nervous that there is only one test case for this PR (reproduced in three different tests).  How confident are you that pre-existing tests cover the other cases that have been discussed?

---

@mathiasvp this work might touch on the upcoming closure captures work, so I think yo should be aware of it.

**Comment by MathiasVP** [comment]
> @MathiasVP this work might touch on the upcoming closure captures work, so I think yo should be aware of it.

Thanks for the heads up. I don't think any of this should impact the dataflow stuff I'm working on, but I'll keep an eye upon.

Do let me know if you need any help from me on this PR. I think the direction of the PR is good, and in general I'm not worried about the implementation strategy Robert has gone with here 👍

File: swift/ql/lib/codeql/swift/elements/expr/ExplicitClosureExpr.qll [hunk 2]
```diff
@@ -1,4 +1,8 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.expr.ExplicitClosureExpr
 
+/**
+ * A Swift explicit closure expr, that is, a closure written using
+ * `{ ... -> ... in ... }` syntax rather than automatically generated by the
+ * compiler .
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/elements/expr/ExplicitClosureExpr.qll:6]:
```suggestion
 * compiler.
```

File: swift/ql/lib/codeql/swift/elements/expr/AutoClosureExpr.qll [hunk 3]
```diff
@@ -4,7 +4,8 @@ private import codeql.swift.elements.expr.Expr
 
 /**
  * A Swift autoclosure expression, that is, a closure automatically generated
- * around an argument when the parameter has the `@autoclosure` attribute. For
+ * around an argument when the parameter has the `@autoclosure` attribute or
+ * for the rtight-hand operand of short-circuiting logical operations. For
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/elements/expr/AutoClosureExpr.qll:8]:
```suggestion
 * for the right-hand operand of short-circuiting logical operations. For
```

**Review by geoffw0 — COMMENTED** [review]
Thanks for adding the QLDoc.

It looks like you need to run `bazel run //swift/codegen`.

I think we're ready for a DCA run as well.

**Pull request merged**